src/community/web-service-auth/src/main/java/org/geoserver/security/auth/web/WebServiceAuthenticationProvider.java

```diff
@@ -0,0 +1,275 @@
+            if (groupAdminRole != null && groupAdminRole.equals(role)) isGroupAdmin = true;
+        }
+        // mark the role as Admin or Group Admin if any of its roles are marked as Admin or Group
+        // Admin
+        if (isAdmin) adminAuthorities.add(GeoServerRole.ADMIN_ROLE);
+        if (isGroupAdmin) adminAuthorities.add(GeoServerRole.GROUP_ADMIN_ROLE);
+        verboseLog(
+                roleService.getName()
+                        + ":User is Admin:"
+                        + isAdmin
+                        + "| User is GroupAdmin: "
+                        + isGroupAdmin);
+        return adminAuthorities;
+    }
+
+    private String encode(String credentials) {
+        return new String(Base64.getEncoder().encode(credentials.getBytes()));
+    }
+
+    private CloseableHttpClient buildHttpClient() {
+        RequestConfig clientConfig =
+                RequestConfig.custom()
+                        .setConnectTimeout(this.config.getConnectionTimeOut() * 1000)
+                        .setSocketTimeout(this.config.getReadTimeoutOut() * 1000)
+                        .build();
+        return HttpClientBuilder.create().setDefaultRequestConfig(clientConfig).build();
+    }
+
+    private HttpGet createGetRequest(Authentication authentication) throws MalformedURLException {
+        String authenticationURL = getAuthenticationURL(config.getConnectionURL(), authentication);
+        HttpGet httpGet = new HttpGet(authenticationURL);
+        if (config.isUseHeader()) addHeaders(httpGet, authentication);
+
+        verboseLog(
+                "External authentication call URL:"
+                        + authenticationURL
+                        + " with headers:"
+                        + Stream.of(httpGet.getAllHeaders())
+                                .collect(Collectors.toMap(Header::getName, Header::getValue)));
+        return httpGet;
+    }
+
+    private boolean isSupportedMimeType(ContentType contentType) {
+        boolean supported = false;
+        if (contentType != null && SUPPORTED_MIME_TYPES.contains(contentType.getMimeType()))
+            supported = true;
+        return supported;
+    }
+
+    private void addRoles(
+            Set<GrantedAuthority> roles, String responseBody, Authentication authentication) {
+        // if a regex is set extract roles from it
+        if (config.getRoleRegex() != null && !config.getRoleRegex().isEmpty()) {
+            roles.addAll(extractRoles(responseBody, config.getRoleRegex()));
+        }
+
+        // next extract user roles from configured service
+        // if no role service is selected, use the system default
+        try {
+            GeoServerRoleService roleService =
+                    getRoleService(authentication.getPrincipal().toString(), config);
+            roles.addAll(authorize(authentication.getPrincipal().toString(), roleService));
+            roles.addAll(checkAdminRoles(roles, roleService));
```

Comment:
Not sure what's happening here, was it already there? If one of the roles is already recognized as admin role, why add it another time?

Reply:
The admin authority is assigned by the the ROLE_ADMIN. The additional check is used to check whether any of the role already returned by the query by username is equal to the admin role known from the roleService if it is then add the ROLE_ADMIN constants. Similar to what happens [here](https://github.com/geoserver/geoserver/blob/de184b2f9ed9a844b5291dd163e772132a07689b/src/main/src/main/java/org/geoserver/security/impl/RoleCalculator.java#L111)